cli/src/feature.rs

```diff
@@ -0,0 +1,273 @@
+        )))
+    }
+}
+
+pub fn parse_feature_subcommand(
+    matches: &ArgMatches<'_>,
+    default_signer: &DefaultSigner,
+    wallet_manager: &mut Option<Arc<RemoteWalletManager>>,
+) -> Result<CliCommandInfo, CliError> {
+    let response = match matches.subcommand() {
+        ("activate", Some(matches)) => {
+            let (feature_signer, feature) = signer_of(matches, "feature", wallet_manager)?;
+            let mut signers = vec![default_signer.signer_from_path(matches, wallet_manager)?];
+            signers.push(feature_signer.unwrap());
+            let feature = feature.unwrap();
+
+            known_feature(&feature)?;
+
+            CliCommandInfo {
+                command: CliCommand::Feature(FeatureCliCommand::Activate { feature }),
+                signers,
+            }
+        }
+        ("status", Some(matches)) => {
+            let mut features = if let Some(features) = pubkeys_of(matches, "features") {
+                for feature in &features {
+                    known_feature(feature)?;
+                }
+                features
+            } else {
+                FEATURE_NAMES.keys().cloned().collect()
+            };
+            features.sort();
+            CliCommandInfo {
+                command: CliCommand::Feature(FeatureCliCommand::Status { features }),
+                signers: vec![],
+            }
+        }
+        _ => unreachable!(),
+    };
+    Ok(response)
+}
+
+pub fn process_feature_subcommand(
+    rpc_client: &RpcClient,
+    config: &CliConfig,
+    feature_subcommand: &FeatureCliCommand,
+) -> ProcessResult {
+    match feature_subcommand {
+        FeatureCliCommand::Status { features } => process_status(rpc_client, features),
+        FeatureCliCommand::Activate { feature } => process_activate(rpc_client, config, *feature),
+    }
+}
+
+// Feature activation is only allowed when 95% of the active stake is on the current feature set
+fn feature_activation_allowed(rpc_client: &RpcClient) -> Result<bool, ClientError> {
+    let my_feature_set = solana_version::Version::default().feature_set;
+
+    let feature_set_map = rpc_client
+        .get_cluster_nodes()?
+        .into_iter()
+        .map(|contact_info| (contact_info.pubkey, contact_info.feature_set))
+        .collect::<HashMap<_, _>>();
+
+    let vote_accounts = rpc_client.get_vote_accounts()?;
+
+    let total_active_stake: u64 = vote_accounts
+        .current
+        .iter()
+        .chain(vote_accounts.delinquent.iter())
+        .map(|vote_account| vote_account.activated_stake)
+        .sum();
+
+    let total_compatible_stake: u64 = vote_accounts
+        .current
+        .iter()
+        .map(|vote_account| {
+            if Some(&Some(my_feature_set)) == feature_set_map.get(&vote_account.node_pubkey) {
+                vote_account.activated_stake
+            } else {
+                0
+            }
+        })
+        .sum();
+
+    Ok(total_compatible_stake * 100 / total_active_stake >= 95)
```

Comment:
I think this is very cool thanks to this runtime feature activation framework.

Also, I don't think this check needs to be on-chain. So, this guard is good to reside on cli here. :)